ffmpeg/api_test.go

```diff
@@ -877,3 +877,79 @@ func consecutiveMuxerOpts(t *testing.T, accel Acceleration) {
 	run(cmd)
 
 }
+
+func TestTranscoder_EncodingProfiles(t *testing.T) {
+	encodingProfiles(t, Software)
+}
+
+func encodingProfiles(t *testing.T, accel Acceleration) {
+	run, dir := setupTest(t)
+	defer os.RemoveAll(dir)
+	cmd := `
+        cp "$1/../transcoder/test.ts" test.ts
+    `
+	run(cmd)
+
+	// First, sanity check defaults
+	for _, p := range VideoProfileLookup {
+		if p.Profile != ProfileNone {
+			t.Error("Default VideoProfile profile not set to ProfileNone")
+		}
+	}
+
+	// Encode to all H264 profiles
+	var profilesMap = map[Profile]string{
+		H264Baseline:        "baseline",
+		H264Main:            "main",
+		H264High:            "high",
+		H264ConstrainedHigh: "constrained_high",
+	}
+	for codecProfile, profileString := range profilesMap {
+		tc := NewTranscoder()
+		profile := P144p30fps16x9
+		profile.Profile = codecProfile
+		in := &TranscodeOptionsIn{Fname: fmt.Sprintf("%s/test.ts", dir), Accel: accel}
+		out := []TranscodeOptions{{
+			Oname:   fmt.Sprintf("%s/out_%s.mp4", dir, profileString),
+			Accel:   accel,
+			Profile: profile,
+		}}
+		_, err := tc.Transcode(in, out)
+		if err != nil {
+			t.Error("Unexpected error ", err)
+		}
+		tc.StopTranscoder()
+	}
+
+	// Verify outputs have correct profile names
+	cmd = `
+		ffprobe -loglevel warning -show_streams out_baseline.mp4 | grep "profile=Constrained Baseline"
+		ffprobe -loglevel warning -show_streams out_main.mp4 | grep "profile=Main"
+		ffprobe -loglevel warning -show_streams out_high.mp4 | grep "profile=High"
+		ffprobe -loglevel warning -show_streams out_constrained_high.mp4 | grep "profile=High"
+	`
+	run(cmd)
+
+	// Verify the two constrained profiles (Baseline & Constrained High) do not have any b frames
```

Comment:
Can we also check that the Main / High profiles do have B-frames in the output? (This might be tricky; IIRC one of the encoders - either x264 or cuda - did not output any B-frames while the other did, under the default configuration.)

Reply:
Yep I had a testcase for that and removed it precisely because NVENC did not encode any bframes even for main/high profile.

I'll add a CPU only test case for it, as I do not want to change the default behaviour of NVENC for main/high profile (even if it is possible to manually force B-frame usage through some parameter)

Reply:
Sounds good - yeah it would help compressibility to incorporate B-frames at some point, but it's not needed for now. Nvidia has some other issues too that feed into this a bit, like it'll output a constant bitrate whether it actually *needs* that many bits. Unlike x264, where the bit rate is more of a maximum, unless you explicitly enable something like cbr or vbv.